### PR TITLE
Refactor generate address 

### PR DESCRIPTION
**Pull request by ManuZiD**
I refactored the generate address logic. Additionally I removed the method org.iota.jota.IotaAPI.getNewAddress this method was already marked as deprecated.

### CODE REVIEW

File: jota/src/test/java/org/iota/jota/IotaAPITest.java [hunk 1]
```diff
@@ -192,15 +255,15 @@ public void shouldPrepareTransfer() throws ArgumentException {
...
     public void shouldPrepareTransferWithInputs() throws ArgumentException {
-        List<Input> inputlist = new ArrayList<>();
+        List<Input> inputList;
         List<Transfer> transfers = new ArrayList<>();
 
         GetBalancesAndFormatResponse rsp = iotaAPI.getInputs(TEST_SEED1, 2, 0, 10, 0);
-        
-        inputlist.addAll(rsp.getInputs());
+
+        inputList  = new ArrayList<>(rsp.getInputs());
```
Comment (kwek20) [inline jota/src/test/java/org/iota/jota/IotaAPITest.java]: nice :) Its just a test so optimization is not that important, but i like it!

File: jota/src/main/java/org/iota/jota/IotaAPI.java [hunk 2]
```diff
@@ -74,186 +77,95 @@ protected IotaAPI(Builder builder) {
...
-    public GetNewAddressResponse generateNewAddresses(String seed, int security, boolean checksum, int amount) throws ArgumentException {
-        return generateNewAddresses(seed, security, checksum, 0, amount, false);
+
+    private boolean isAddressSpent(String newAddress, boolean checksum) {
+        final String address = checksum ? newAddress : Checksum.addChecksum(newAddress);
+        final FindTransactionResponse response = findTransactionsByAddresses(address);
+
+        if (response.getHashes().length == 0) {
+            WereAddressesSpentFromResponse spent = wereAddressesSpentFrom(address);
+            return spent.getStates().length > 0 && !spent.getStates()[0];
```
Comment (kwek20) [inline jota/src/main/java/org/iota/jota/IotaAPI.java]: You can change this to `checkWereAddressSpentFrom` (And maybe rename it to checkWas...)

File: jota/src/main/java/org/iota/jota/IotaAPI.java [hunk 3]
```diff
@@ -74,186 +77,95 @@ protected IotaAPI(Builder builder) {
...
+        }
+
+        return false;
```
Comment (kwek20) [inline jota/src/main/java/org/iota/jota/IotaAPI.java]: if a transaction has hashes attached, we should check if they were signed ones before returning false. 0 value ones without signing are okay

File: jota/src/main/java/org/iota/jota/IotaAPI.java [hunk 4]
```diff
@@ -74,186 +77,95 @@ protected IotaAPI(Builder builder) {
...
+    public GetNewAddressResponse getAddressesUnchecked(AddressRequest addressRequest) throws ArgumentException {
         StopWatch stopWatch = new StopWatch();
 
-        List<String> allAddresses = new ArrayList<>();
-        for (int i = index; i < index + amount; i++) {
-            allAddresses.add(IotaAPIUtils.newAddress(seed, security, i, checksum, getCurl()));
-        }
+        List<String> allAddresses = IntStream
+                .range(addressRequest.getIndex(), addressRequest.getIndex() + addressRequest.getAmount())
+                .mapToObj(mIndex -> IotaAPIUtils.newAddress(addressRequest, mIndex, getCurl()))
```
Comment (kwek20) [inline jota/src/main/java/org/iota/jota/IotaAPI.java]: :)

File: jota/src/main/java/org/iota/jota/builder/AddressRequest.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        private int index = 0;
+        private boolean checksum = false;
+        private boolean addSpendAddresses = false;
+
+        public Builder(String seed, int securityLevel) {
+            this.seed = InputValidator.requireValidSeed(seed);
+            this.securityLevel = InputValidator.requireValidSecurityLevel(securityLevel);
+        }
+
+        public Builder amount(int amount) {
```
Comment (kwek20) [inline jota/src/main/java/org/iota/jota/builder/AddressRequest.java]: Can you add the comment about 0 amount being the first, and less than 0 should throw an error or count backwards. What do you think about that?

File: jota/src/test/java/org/iota/jota/IotaAPITest.java (hunk 1 above)
Comment (ManuZiD) [reply jota/src/test/java/org/iota/jota/IotaAPITest.java]: Yes, after the push I saw it on several other places. I think we can re-organise the tests in an other PR.

File: jota/src/main/java/org/iota/jota/IotaAPI.java (hunk 3 above)
Comment (kwek20) [reply jota/src/main/java/org/iota/jota/IotaAPI.java]: I actually havent tested this; was just something that came to mind when reviewing.
But you should be able to use `utils/Signing.validateSignatures`

Low priority though, we can also make a ticket and do something more directly impacting

**Pull request merged**